### ISSUE TITLE
Using proto_source_root in proto_library breaks java_grpc_library

### ISSUE THREAD

**Issue by haikalpribadi** [grpc/grpc-java#4896]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?

v1.15.0

### What did you expect to see?

Say I have 2 proto files sitting in `//some-package/`
- `//some-package/A.proto` which `imports B.proto` (without `some-package` as prefix)
- `//some-package/B.proto`

Then I compile the `proto_library in` //some-package/BUILD`, like the following:
```
proto_library(
    name = "some-proto",
    srcs = glob(["*.proto"])
    proto_source_root = "some-package"
)
```
The above works well because `proto_source_root`, which allows us to not have to import one proto file into another using the full path relative from WORKSPACE.

As instructed on GRPC docs (https://grpc.io/docs/reference/java/generated-code.html) Next I compile `java_proto_library`:
```
java_proto_library(
    name = "java-proto",
    deps = ["some-proto"],
)
```
This also works fine. But finally, I compile `java_grpc_library`:
```
java_grpc_library(
    name = "java-grpc",
    srcs = [":some-proto"],
    deps = [":java-proto"],
)
```
And here it breaks. `java_grpc_library` failes to compile with error:
```
B.proto: file not found.
some-package/A.proto: Import "B.proto" was not found or had errors.
```
So basically, `A.proto` cannot find `B.proto`. However, when I replace the import statement in `A.proto` with `import some-package/B.proto` and remove usage of `proto_source_root` in `proto_library` everything works, but we really don't want to do this for many reasons.

Does anyone know what could be the cause and how to work around this?

**Comment by haikalpribadi** [comment]
To help reproduce this problem easily, I've isolated the problem on a branch on my repo: https://github.com/haikalpribadi/grakn/tree/bazel-init-grpc-java-broken-with-proto-source-root

Thank you!

**Comment by haikalpribadi** [comment]
Thans, @ejona86. Is there any examples on how to use file/transitive_descriptor_sets ?

Although, I do think adding `proto_source_root` to `java_grpc_library` (and every other `*_grpc_library` for that matter) is a more elegant solution. And an even better solution is for `java_grpc_library` to not need `proto_source_root` at all, and instead rely on the fact that the `proto_library` has already encapsulated this information. Is that possible?

**Comment by ejona86** [comment]
> And an even better solution is for java_grpc_library to not need proto_source_root at all, and instead rely on the fact that the proto_library has already encapsulated this information.

This is what I was suggesting with the descriptor set.

> Although, I do think adding proto_source_root to java_grpc_library (and every other *_grpc_library for that matter) is a more elegant solution.

For the solution like this, java_grpc_library wouldn't need any additional configuration, it would just be changed to look at proto_library's configuration.

**Comment by haikalpribadi** [comment]
> This is what I was suggesting with the descriptor set.

I see! Sorry I didn't get your first point, but I do now. :)

> For the solution like this, java_grpc_library wouldn't need any additional configuration, it would just be changed to look at proto_library's configuration.

Do you mean `java_grpc_library` would still need a new argument but it will have a default value which is what is inside `proto_library`'s configuration?

**Comment by ejona86** [comment]
> Do you mean java_grpc_library would still need a new argument but it will have a default value which is what is inside proto_library's configuration?

No, I was meaning it would use the value directly from the proto_library. Although we're not using aspects in java_grpc_library, that's more of the style of how proto_library is expected to be integrated with.

For reference, here's the type of changes necessary for using the descriptor set. But it is based on the internal copy which is basically slowly being rewritten and I've not yet exported the changes (limited time and I'm "not done yet").

```diff
--- a/java/grpc/build_defs.bzl	2018-07-23 17:24:35.000000000 -0700
+++ b/java/grpc/build_defs.bzl	2018-07-25 13:48:29.000000000 -0700
@@ -10,9 +10,6 @@
 

 
-def _create_include_path(include):
-    return "-I{0}={1}".format(include.short_path, include.path)
-
 def _short_path(src):
     return src.short_path
 
@@ -24,7 +21,7 @@
                "same package as consuming rule").format(ctx.label, ctx.attr.srcs[0].label))
 
     srcs = ctx.attr.srcs[0].proto.direct_sources
-    includes = ctx.attr.srcs[0].proto.transitive_imports
+    descriptor_set_in = ctx.attr.srcs[0].proto.transitive_descriptor_sets
     flavor = ctx.attr.flavor
     if flavor == "normal":
         flavor = ""
@@ -32,11 +29,11 @@
     args = ctx.actions.args()
     args.add(ctx.executable._java_plugin.path, format = "--plugin=protoc-gen-grpc-java=%s")
     args.add("--grpc-java_out={0}:{1}".format(flavor, ctx.outputs.srcjar.path))
-    args.add_all(includes, map_each = _create_include_path)
+    args.add_joined("--descriptor_set_in", descriptor_set_in, join_with = ":")
     args.add_all(srcs, map_each = _short_path)
 
     ctx.action(
-        inputs = depset([ctx.executable._java_plugin] + srcs, transitive = [includes]),
+        inputs = depset([ctx.executable._java_plugin] + srcs, transitive = [descriptor_set_in]),
         outputs = [ctx.outputs.srcjar],
         executable = ctx.executable._protoc,
         arguments = [args],
```

**Comment by haikalpribadi** [comment]
Alright, I see. It makes to me from a high level, but unfortunately, my 2 weeks old knowledge of Bazel might not be enough for me to submit a PR for this. :/

**Comment by adibrastegarnia** [comment]
Hello, 

I have a similar problem but I didn't understand the solution. I used proto source root in my proto_library and it can be compiled fine but when I use grpc_proto_library to generate grpc stubs, it gives import not found and file not found (i.e. it cannot fine the imports in the protobuf that I define the services)

**Comment by adibrastegarnia** [comment]
@ejona86 

Eric, 

May I ask you what is the status of this change? Can you give me some guidelines If I want to work on it. We need to get it fixed to use for a project. 

Thanks,
Adib

**Comment by ejona86** [comment]
Honestly, for the while my recommendation is "don't do that;" always import from the base directory. This is what Blaze has required of us, for example. So I see this more as a feature request with a possible workaround.

Why is this fix _required_ for a project?

**Comment by adibrastegarnia** [comment]
@ejona86 

Let me explain our situation: Please take a look at: 
[(https://github.com/dnosproject/dnos-core-grpc/blob/master/protobuf/proto/net/ServicesProto.proto)]
we use imports like this for example : import "net/link/LinkEvent.proto" if we want to import from LinkEvent.proto in another protobuf message. It works fine if we don't use grpc_proto_lib  to generate a grpc stub. If we want to use grpc_proto_lib then the imports should be changed to something like this otherwise it cannot find it (i.e. file not found error): import "protobuf/proto/net/link/LinkEventProto.proto". If you look at the build file we use proto source root in proto_build rules.  

https://github.com/dnosproject/dnos-core-grpc/blob/master/protobuf/proto/BUILD

Does that mean we should change our imports or can we still use net/link/LinkEvent.proto?

**Comment by adibrastegarnia** [comment]
@ejona86 
OK. Thanks for letting me know.  Please keep me posted if you find a solution (I hope you do).

**Comment by adibrastegarnia** [comment]
@ejona86 

Eric, Any updates on this issue?

**Comment by ejona86** [comment]
No, no updates.

**Comment by ejona86** [comment]
I dug into this more. I may be missing something simple, but it currently seems mostly unsupportable. It seems like it requires using an aspect, plus some very specific logic. I tried using the transitive descriptors, and they appear not to save any effort with this particular feature.

The only ways this could be implemented would be for ProtoInfo to include more information or for proto_toolchain to open itself up to Skylark implementations.

To give a sampling of how Bazel itself is using this stuff:
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCommon.java#L288
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCompileActionBuilder.java#L698

Part of the problem is that when using proto_source_root you can still do imports based on the full path. (That is, it seems you can take any proto_library and add proto_source_root to it without breaking the build.)

I'll need to have some conversations with the Bazel folks at some point.

**Comment by adibrastegarnia** [comment]
Thanks Eric for the update.

> Part of the problem is that when using proto_source_root you can still do
imports based on the full path. (That is, it seems you can take any
proto_library and add proto_source_root to it without breaking the build.)

This is exactly what I do as a workaround. Do you think should we go for
that workaround or wait to get it fixed?

Adib

**Comment by ejona86** [comment]
@adibrastegarnia, I don't really see how that is a "workaround" as it is equivalent to not using proto_source_root. For now, I'd recommend not using proto_source_root. I'm not expecting a quick fix. But I also need to talk with the Bazel folks.

**Comment by adibrastegarnia** [comment]
@ejona86 
OK. Please let me know if there is any updates.

**Comment by ejona86** [comment]
FYI: https://github.com/bazelbuild/bazel/issues/7153 , `proto_source_root` is being replaced with `strip_import_prefix`

**Comment by adibrastegarnia** [comment]
Thanks Eric for the update. Glad to hear it gets fixed.

**Comment by ejona86** [comment]
Fixed by #5959
